dissect/target/tools/query.py

```diff
@@ -110,66 +154,125 @@ def main():
+            plugin_target = open_target(target)
+            functions = find_plugin_functions(plugin_target, args.function, True)
+            tasklist.append((plugin_target, functions))
+        tasks = tasks_opened(tasklist)
+    else:
+        plugin_target = Target()
+        functions = find_plugin_functions(plugin_target, args.function, False)
+        tasklist.append((plugin_target, functions))
+        if args.targets:
+            tasks = tasks_unopened(args.targets, args.children, args.child, functions)
+        else:
+            tasks = tasks_opened(tasklist)
 
-    # Show help for a function
-    if "-h" in rest or "--help" in rest:
-        first_func = functions[0]
-        obj = get_attr_for_attr_path(first_func)
+    # Show the list of available plugins for the given optional target and optional
+    # search pattern, only display plugins that can be applied to ANY targets
+    if args.list:
+        collected_plugins = {}
+        for index, task in enumerate(tasklist):
+            _, funcs = task
+            for func in funcs:
+                collected_plugins[func.name] = func.plugin_desc
+
+        # Display in a user friendly manner
+        target = Target()
+        fparser = generate_argparse_for_bound_method(target.plugins, usage_tmpl=USAGE_FORMAT_TMPL)
+        fargs, rest = fparser.parse_known_args(rest)
+        target.plugins(list(collected_plugins.values()))
+        # No real targets specified, user might be interested to see available loaders
+        if not args.targets:
+            fparser = generate_argparse_for_bound_method(target.loaders, usage_tmpl=USAGE_FORMAT_TMPL)
+            fargs, rest = fparser.parse_known_args(rest)
+            target.loaders(**vars(fargs))
+        parser.exit()
 
-        if not obj:
-            parser.error(f"Can't find plugin with function `{first_func}`")
+    # If we get here, that means we need to execute the collected plugins,
+    # first check if the output types are uniform, if not:
+    # - emit warning
+    # - fall back to record and
+    # - ignore further errors of this kind
+    total_num_tasks = 0
+    output_types = set()
+
+    for task in tasklist:
+        _, funcs = task
+        for func in funcs:
+            total_num_tasks += 1
+            output_types.add(func.output_type)
+    # Important! By setting the default output to record we also
+    # indicate that further errors because of mixed output types will
+    # be supressed (makes no sense anymore, we already have set the default)
+    default_output_type = None
+    if len(output_types) > 1:
+        log.warning(f"Mixed output types detected: {'/'.join(output_types)}, only outputting records.")
+        default_output_type = "record"
 
+    # Show help for a function
+    if "-h" in rest or "--help" in rest:
+        if not total_num_tasks:
+            parser.error("Function(s) not found, see -l for available plugins")
+            parser.exit()
+        _, func_list = list(tasks())[0]
```

Comment:
Probably the only unfortunate thing here is that when doing a `target-query -f browser -h <TARGET>` the TARGET will be opened although it is not needed.

Reply:
Fixed